app/Components/Game/index.js

```diff
@@ -45,28 +45,12 @@ class Game extends React.Component {
     this.setState({roundBegan: true});
   }
 
-  showRoleOptions(takenRoles) {
-    var studentDisabled = false;
-    var teacherDisabled = false;
-    if (takenRoles.length == 2) {
-      studentDisabled = true;
-      teacherDisabled = true;
-    } else if (takenRoles.length == 1) {
-      if (takenRoles[0] == "student") {
-        studentDisabled = true;
-      } else {
-        teacherDisabled = true;
-      }
-    }
-    return(
-      <RoleSelectionMenu 
-        socket={socket} 
-        takenRoles={this.state.takenRoles} 
-        selectRole={this.selectRole}
-        gameID={this.props.params.gameID}
-        studentDisabled={studentDisabled}
-        teacherDisabled={teacherDisabled}/>
-    );
+  checkTakenRoles(callback) {
+    socket.emit('checkTakenRoles', this.props.params.gameID, (takenRoles) => {
+      this.setState({takenRoles: takenRoles});
+      callback(this.state.takenRoles);
```

Comment:
This looks good, but this sets up a chain of event handlers that are then decoupled from the lifecycle of the React components.

For example:
1. a child component calls `checkTakenRoles`
2. it emits an event on the socket
3. ...time passes while the network communication happens...
4. ...let's say the user navigates away to start a new game and this component unmounts and is destroyed...
5. the network communication finishes and the callback here fires
6. this component tries to `setState` and a warning is triggered since it's not mounted on the page anymore
7. this component calls `callback` and then the child component also tries to update itself and triggers another warning since it's not mounted on the page anymore either.

There's some more background in the [React docs](https://facebook.github.io/react/blog/2015/12/16/ismounted-antipattern.html) on this kind of sequence if you're curious.

Concretely, one suggestion could be to have this method only `emit`, and not add a new listener for the server response.  That would let you move all the event listening code (eg., for when `checkTakenRoles` comes back from the server) into the `componentDidMount` method, and then remove all of them in `componentDidUnmount`.  This will probably take a bit of refactoring, but also may help simplify some of the chains of callbacks, so the data flow is a bit more explicit, where either:

- a UI event happened, and that leads to calling setState and/or emitting a server event
- a server event came in, and that leads to calling setState

Also, I'm happy to chat about this or pair on it at some point, I think it'd be a good improvement to make but certainly isn't critical to do it immediately, or to rewrite all the things before the playtest. :)